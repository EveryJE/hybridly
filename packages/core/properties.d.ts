type Property = null | string | number | boolean | Property[] | { [name: string]: Property }
type Properties = Record<string | number, Property>

interface GlobalHybridlyProperties {
	[key: string | number]: Property
}
