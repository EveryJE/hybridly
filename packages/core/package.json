{
  "name": "@monolikit/core",
  "version": "0.1.0",
  "description": "A solution to develop server-driven, client-rendered applications",
  "keywords": [
    "monolikit",
    "inertiajs"
  ],
  "homepage": "https://github.com/monolikit/monolikit#readme",
  "bugs": {
    "url": "https://github.com/monolikit/monolikit/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/monolikit/monolikit.git"
  },
  "funding": "https://github.com/sponsors/innocenzi",
  "license": "MIT",
  "author": "Enzo Innocenzi <enzo@innocenzi.dev>",
  "sideEffects": false,
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    }
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
		"*.d.ts"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "dependencies": {
    "debug": "^4.3.4",
    "deepmerge": "^4.2.2",
    "hookable": "^5.1.1",
    "qs": "^6.11.0"
  },
  "peerDependencies": {
    "axios": "^0.27"
  },
  "devDependencies": {
    "axios": "^0.27.2",
    "defu": "^6.0.0"
  }
}
