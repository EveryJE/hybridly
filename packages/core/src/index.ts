export { createRouter, router } from './router'
export type { Router, VisitOptions, VisitResponse, VisitPayload, ResolveComponent, Method } from './router'

export { getRouterContext } from './context'
export type { RouterContext, RouterContextOptions } from './context'

export { definePlugin, registerHook, registerHookOnce } from './plugins'
export type { Plugin } from './plugins'

export { makeUrl, sameUrls } from './url'
export type { UrlResolvable } from './url'

export { can } from './authorization'
export type { Authorizable } from './authorization'

export * as constants from './constants'
export * from './types'
