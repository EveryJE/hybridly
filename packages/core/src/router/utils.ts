export const STORAGE_EXTERNAL_KEY = 'sleightful:external'

/**
 * Checks if the current visit was made by going back or forward.
 */
export function isBackForwardVisit(): boolean {
	if (!window.history.state) {
		return false
	}

	return (window.performance?.getEntriesByType('navigation').at(0) as PerformanceNavigationTiming)?.type === 'back_forward'
}

/**
 * Checks if the current visist points to an external location.
 */
export function isExternalVisit(): boolean {
	try {
		return window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) !== null
	} catch {}

	return false
}

/**
 * Gets the current visit type.
 */
export function getCurrentVisitType(): 'back_forward' | 'external' | 'sleightful' {
	if (isBackForwardVisit()) {
		return 'back_forward'
	}

	if (isExternalVisit()) {
		return 'external'
	}

	return 'sleightful'
}

/**
 * Convers the given string or URL to an URL.
 */
export function makeUrl(href: string | URL): URL {
	return new URL(href.toString(), window.location.toString())
}
