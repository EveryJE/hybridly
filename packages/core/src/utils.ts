import makeDebugger from 'debug'

export const debug = {
	router: makeDebugger('sleightful:router'),
	history: makeDebugger('sleightful:history'),
	url: makeDebugger('sleightful:url'),
	context: makeDebugger('sleightful:context'),
	external: makeDebugger('sleightful:external'),
	scroll: makeDebugger('sleightful:scroll'),
	event: makeDebugger('sleightful:event'),
	adapter: (name: string, ...args: any[]) => makeDebugger('sleightful:adapter').extend(name)(args.shift(), ...args),
}

export function match<TValue extends string | number = string, TReturnValue = unknown>(
	value: TValue,
	lookup: Record<TValue | 'default', TReturnValue | ((...args: any[]) => TReturnValue)>,
	...args: any[]
): TReturnValue | Promise<TReturnValue> {
	if (value in lookup || 'default' in lookup) {
		const returnValue = value in lookup
			? lookup[value]
			: lookup.default

		return typeof returnValue === 'function' ? returnValue(...args) : returnValue
	}

	const handlers = Object.keys(lookup)
		.map((key) => `"${key}"`)
		.join(', ')

	const error = new Error(`Tried to handle "${value}" but there is no handler defined. Only defined handlers are: ${handlers}.`)

	if (Error.captureStackTrace) {
		Error.captureStackTrace(error, match)
	}

	throw error
}

export function debounce<F extends(...params: any[]) => ReturnType<F>>(fn: F, delay: number): F {
	let timeoutID: NodeJS.Timeout
	return function(...args: unknown[]) {
		clearTimeout(timeoutID)
		timeoutID = setTimeout(() => fn(args), delay)
	} as F
}

export function value<T>(value: T | (() => T)): T {
	if (typeof value === 'function') {
		return (value as any)?.() as T
	}

	return value
}

export function when<T, D>(condition: any, data: T, _default?: D): T | D | undefined {
	if (!condition) {
		return _default
	}

	return data
}
