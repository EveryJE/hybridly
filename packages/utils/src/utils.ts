import baseMerge from 'deepmerge'

export function random(length: number = 10): string {
	const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

	let str = ''
	for (let i = 0; i < length; i++) {
		str += chars.charAt(Math.floor(Math.random() * chars.length))
	}

	return str
}

export function match<TValue extends string | number = string, TReturnValue = unknown>(
	value: TValue,
	lookup: Record<TValue | 'default', TReturnValue | ((...args: any[]) => TReturnValue)>,
	...args: any[]
): TReturnValue | Promise<TReturnValue> {
	if (value in lookup || 'default' in lookup) {
		const returnValue = value in lookup
			? lookup[value]
			: lookup.default

		return typeof returnValue === 'function' ? returnValue(...args) : returnValue
	}

	const handlers = Object.keys(lookup)
		.map((key) => `"${key}"`)
		.join(', ')

	const error = new Error(`Tried to handle "${value}" but there is no handler defined. Only defined handlers are: ${handlers}.`)

	throw error
}

export function debounce<F extends(...params: any[]) => ReturnType<F>>(fn: F, delay: number): F {
	let timeoutID: ReturnType<typeof setTimeout>
	return function(...args: unknown[]) {
		clearTimeout(timeoutID)
		timeoutID = setTimeout(() => fn(args), delay)
	} as F
}

export function value<T>(value: T | (() => T)): T {
	if (typeof value === 'function') {
		return (value as any)?.() as T
	}

	return value
}

export function when<T, D>(condition: any, data: T, _default?: D): T | D | undefined {
	if (!condition) {
		return _default
	}

	return data
}

export function clone<T>(val: T): T {
	// Thanks Anthony <3
	let k: any, out: any, tmp: any

	if (Array.isArray(val)) {
		out = Array(k = val.length)
		while (k--) {
			// eslint-disable-next-line no-cond-assign
			out[k] = (tmp = val[k]) && typeof tmp === 'object' ? clone(tmp) : tmp
		}

		return out as any
	}

	if (Object.prototype.toString.call(val) === '[object Object]') {
		out = {} // null
		// eslint-disable-next-line no-restricted-syntax
		for (k in val) {
			if (k === '__proto__') {
				Object.defineProperty(out, k, {
					value: clone((val as any)[k]),
					configurable: true,
					enumerable: true,
					writable: true,
				})
			} else {
				// eslint-disable-next-line no-cond-assign
				out[k] = (tmp = (val as any)[k]) && typeof tmp === 'object' ? clone(tmp) : tmp
			}
		}

		return out
	}

	return val
}

export function merge<T>(x: Partial<T>, y: Partial<T>): T {
	return baseMerge(x, y, { arrayMerge: (_, s) => s })
}
