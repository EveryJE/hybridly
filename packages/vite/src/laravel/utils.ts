import type { AddressInfo } from 'node:net'
import { fileURLToPath } from 'node:url'

export function dirname(): string {
	return fileURLToPath(new URL('.', import.meta.url))
}

export function isIpv6(address: AddressInfo): boolean {
	return address.family === 'IPv6'
	// In node >=18.0 <18.4 this was an integer value. This was changed in a minor version.
	// See: https://github.com/laravel/vite-plugin/issues/103
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-expect-error-next-line
		|| address.family === 6
}
