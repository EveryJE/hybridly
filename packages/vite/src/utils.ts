import makeDebugger from 'debug'
import { LAYOUT_PLUGIN_NAME, ROUTING_PLUGIN_NAME } from './constants'

export const debug = {
	router: makeDebugger(ROUTING_PLUGIN_NAME),
	layout: makeDebugger(LAYOUT_PLUGIN_NAME),
}
