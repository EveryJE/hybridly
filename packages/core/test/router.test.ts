// @vitest-environment happy-dom
import { beforeEach, expect, test, vi } from 'vitest'
import { initialize } from '@sleightful/core'
import { fakeRouterOptions, fakeVisit, returnsArgs } from './utils'

beforeEach(() => {
	window.location.href = 'https://localhost'
})

test('it performs a visit when initializing', async() => {
	const spy = vi.spyOn(window.history, 'pushState')

	await initialize(fakeRouterOptions())

	expect(spy).toHaveBeenCalledTimes(1)
})

test('it saves visits in the history state upon navigation', async() => {
	const spy = vi.spyOn(window.history, 'pushState').mockImplementation(returnsArgs)

	const router = await initialize(fakeRouterOptions())
	await router.navigate({
		visit: fakeVisit({
			view: {
				name: 'another-view',
				url: 'https://localhost/another-url',
			},
		}),
	})

	expect(spy).toHaveBeenCalledTimes(2)
	expect(spy.mock.results).toMatchSnapshot()
})
