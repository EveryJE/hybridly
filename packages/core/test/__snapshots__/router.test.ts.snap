// Vitest Snapshot v1

exports[`it saves visits in the history state upon navigation 1`] = `
[
  {
    "type": "return",
    "value": [
      {
        "context": "",
        "type": "page",
        "version": "",
        "view": {
          "name": "my.component",
          "properties": {},
          "url": "http://localhost",
        },
      },
      "",
      "http://localhost",
    ],
  },
  {
    "type": "return",
    "value": [
      {
        "context": "",
        "type": "page",
        "version": "",
        "view": {
          "name": "another-view",
          "url": "https://localhost/another-url",
        },
      },
      "",
      "https://localhost/another-url",
    ],
  },
]
`;
